harness/determined/common/experimental/determined.py

```diff
@@ -54,33 +54,24 @@ def __init__(
             explicit_noverify=noverify,
         )
 
-        auth = authentication.Authentication(self._master, user, password, cert=cert)
+        utp = authentication.login_with_cache(self._master, user, password, cert=cert)
```

Comment:
I like it when constructors in Python can't fail (though I can't remember, right now, why I like it). What do you think about replacing this default constructor and `_from_session` with a default constructor that takes a `_session` and that's it?

`client` is class with the convenient interface, anyway. Might as well make this one simple?

Reply:
I like it when constructors only save their parameters to `self` and have little or no additional logic.

I like it because when you put any sort of complex logic into the contstructor, it becomes hard or impossible to create that object via another means.  See the `LogoutAuthentication` as one of the worst offenders of that principle.

I am not in support of a breaking API change as part of this PR.  If you want to break the `client.Determined` constructor API, that's fine, but it won't be part of this PR.

It would be an improvement though.  Basically, the only reason we can even implement `_from_session()` after writing this object with a "smart" constructor, is because python so graciously provides us a way to call `__new__` without calling `__init__`.

(thank you python, for saving us from our own mistakes).

Side note... I would support using a `.from_*()` classmethod as the public-facing way to create a Determined object, but `from_session()` should not be the public one, since the Session itself is not a public api.  It would be more like `client.Determined.from_login()` or even a standalone helper function like `client.new_determined(master=..., user=..., password=...)`.  (obviously not called `new_determined()` because that's an awful name, but you get the idea).

Eh, now that I think about it, from a user perspective, the way the `Determined()` constructor works is fine.  I don't like the pattern in general but I am not sure if I'd be willing to support a breaking change that resulted in a more awkward user-facing API, even if internally the implementation was more principled.

Reply:
Thanks for reminding me why I like those same simple, passthrough constructors in Python.

Mmm. I think I'd maybe forgotten that `Determined` was public.

I think we're agreed that the set of parameters the default constructor takes is the only way a user should get themselves a `Determined`, and that if the associated login fails, the user should not be able to create an object.

Resolving.